src/Components/Components/src/Rendering/Renderer.cs

```diff
@@ -57,24 +48,36 @@ public abstract class Renderer : IDisposable
         /// <param name="serviceProvider">The <see cref="IServiceProvider"/> to be used when initializing components.</param>
         public Renderer(IServiceProvider serviceProvider)
         {
+            if (serviceProvider is null)
+            {
+                throw new ArgumentNullException(nameof(serviceProvider));
+            }
+
             _componentFactory = new ComponentFactory(serviceProvider);
+            Dispatcher = NullDispatcher.Instance;
         }
 
         /// <summary>
         /// Constructs an instance of <see cref="Renderer"/>.
         /// </summary>
         /// <param name="serviceProvider">The <see cref="IServiceProvider"/> to be used when initializing components.</param>
-        /// <param name="dispatcher">The <see cref="IDispatcher"/> to be for invoking user actions into the <see cref="Renderer"/> context.</param>
-        public Renderer(IServiceProvider serviceProvider, IDispatcher dispatcher) : this(serviceProvider)
+        /// <param name="dispatcher">The <see cref="Dispatcher"/> to be for invoking user actions into the <see cref="Renderer"/> context.</param>
+        public Renderer(IServiceProvider serviceProvider, Dispatcher dispatcher)
+            : this(serviceProvider)
         {
-            _dispatcher = dispatcher;
+            Dispatcher = dispatcher ?? NullDispatcher.Instance;
         }
 
         /// <summary>
-        /// Creates an <see cref="IDispatcher"/> that can be used with one or more <see cref="Renderer"/>.
+        /// Gets the <see cref="Microsoft.AspNetCore.Components.Dispatcher" /> associated with this <see cref="Renderer" />.
         /// </summary>
-        /// <returns>The <see cref="IDispatcher"/>.</returns>
-        public static IDispatcher CreateDefaultDispatcher() => new RendererSynchronizationContext();
+        public Dispatcher Dispatcher { get; }
+
+        /// <summary>
+        /// Creates an <see cref="Dispatcher"/> that can be used with one or more <see cref="Renderer"/>.
+        /// </summary> 
+        /// <returns>The <see cref="Dispatcher"/>.</returns>
+        public static Dispatcher CreateDefaultDispatcher() => new RendererSynchronizationContextDispatcher();
```

Comment:
This pattern of having a static `CreateDefaultDispatcher` plus an optional constructor param seems unusual to me, so I'd be interested to understand the goals.

I would have found it more intuitive to have something like an abstract protected property for it, so that it's not needed as a constructor param, and you can't get into a situation where the wrong type of dispatcher is used with a particular type of renderer.

Reply:
OK. I erred on the side of leaving this intact. I'll clean it up.

I think there's a problem with your suggestion, we don't want to require different base-classes of the renderer just to use a different dispatcher type.

Reply:
> we don't want to require different base-classes of the renderer just to use a different dispatcher type

What's the use case for varying which dispatcher is used with which renderer?

I was thinking that tying these together was a benefit, as it's one fewer thing that can go wrong, and I couldn't think of any use case for the association not being known at compile-time.

Reply:
Is it just unit tests?

Reply:
Does this discussion get easier if the next thing we do is make `Renderer` internal? I don't think we've identified any use cases for the rendering infrastructure in this release, and I'm sure it's going to change when we try to solve testing.

Reply:
I don't think `Renderer` could be internal without killing things like Blectron

Reply:
I went down the path of having the `Dispatcher` property as abstract and I'm not sure that the result is totally worse or better.  It gets rid of the weird `CreateDefault` method so that's a plus.

One concrete problem this causes is that if you need the dispatcher for anything before the renderer is created, you're just out of luck. I can't think of why that would be, but making the dispatcher a detail of the renderer will do that.

You can see the result of this here: 21535cd8d14580eef00ec677104720d3f014ecd0

I'm going to do the opposite and see what happens if we just make it a normal constructor parameter.

Reply:
> I'm going to do the opposite and see what happens if we just make it a normal constructor parameter.

This btw is the reason why I'm looking for a way to make the renderer internal, or possibly only internally constructable. What happens to **public constructable infrastructure types** like the renderer is that the set of DI services they depend upon will grow over the next three releases, which leaves a minefield of constructors with `[Obsolete]`.

Reply:
Ok that's not great either. I'm going update this PR with the idea that you had, and we can pick this up again when we get to reviewing the renderer.